pdbeccdutils/core/boundmolecule.py

```diff
@@ -0,0 +1,313 @@
+#!/usr/bin/env python
+# software from PDBe: Protein Data Bank in Europe; https://pdbe.org
+#
+# Copyright 2018 EMBL - European Bioinformatics Institute
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on
+# an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied. See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+"""
+A set of methods for identifying bound-molecules (covalently bonded CCDs ) from mmCIF files
+of proteins and creating MultiDiGraph representation of the molecules.
+"""
+
+from gemmi import cif
+from pdbeccdutils.helpers import helper
+from networkx import MultiDiGraph, connected_components
+from pdbeccdutils.core.models import (
+    Residue,
+    AssemblyResidue,
+    BoundMolecule,
+)
+
+
+def infer_bound_molecules(structure, to_discard, assembly=False):
+    """Identify bound molecules in the input protein structure.
+
+    Args:
+        structure (str): Path to the structure.
+        to_discard (list of str): List of residue names to be discarded
+    """
+
+    bms = parse_bound_molecules(structure, to_discard, assembly)
+    bound_molecules = []
+
+    for bm_nodes in connected_components(bms.to_undirected()):
+        subgraph = bms.subgraph(bm_nodes)
+        bm = BoundMolecule(subgraph)
+        bound_molecules.append(bm)
+
+    bound_molecules = sorted(bound_molecules, key=lambda l: -len(l.graph.nodes))
+    return bound_molecules
+
+
+def __add_connections(struct_conn: dict[str, list[str]], bms: MultiDiGraph) -> None:
+    """Adds covalent connections between chemical components in a
+    bound-molecule from `_struct_conn` table
+
+    Args:
+        struct_conn: Dictionary of `_struct_conn` category from mmCIF
+        bms: Graph corresponding to bound-molecule
+    """
+    for i in range(len(struct_conn["id"])):
+        (ptnr1, atom1) = find_pntr_entry(struct_conn, bms.nodes, 1, i)
+        (ptnr2, atom2) = find_pntr_entry(struct_conn, bms.nodes, 2, i)
+
+        # we want covalent connections among ligands only.
+        if ptnr1 and ptnr2 and struct_conn["conn_type_id"][i] != "metalc":
+            add_edge = True
+            # In MultiDiGraph two nodes can have multiple edges even if the edges share the same edge data.
+            # _struct_conn has duplicate links between atoms, they are removed by the below process.
+            edge_data = bms.get_edge_data(ptnr1, ptnr2)
+            if edge_data:
+                for _, value in edge_data.items():
+                    if value["atom_id_1"] == atom1 and value["atom_id_2"] == atom2:
+                        add_edge = False
```

Comment:
No. Since it is a MultiDiGraph two nodes can have multiple edges with same/different edge data. here, the edge data is atoms connected beetween two nodes. This step is to avoid duplicate edges (same nodes and atoms). When add_edge becomes False, it means the connection between those atoms are already added, but we still need to check for connection through other atoms between the same nodes